### PR TITLE
Add Non Retriable Error Wrapper for all SDK's

### PR DESCRIPTION
**Pull request by MBSolomon**
This PR adds a Non Retriable Error Wrapper to the azure-sdk-for-go repository. #19467 

[Azure/autorest.go]: https://github.com/Azure/autorest.go

### CODE REVIEW

File: sdk/internal/shared/shared.go [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+// Licensed under the MIT License.
+
+package shared
+
+import (
+	"github.com/Azure/azure-sdk-for-go/sdk/internal/errorinfo"
+)
+
+// NonRetriableError marks the specified error as non-retriable.
+func NonRetriableError(err error) error {
```
Comment (jhendrixMSFT) [inline sdk/internal/shared/shared.go:14]: Might simplify things to move this to the `errorinfo` package so this and `NonRetriable` are together.

File: sdk/internal/errorinfo/errorinfo_test.go [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+
+package errorinfo
+
+import (
+	"testing"
+)
+
+func TestNonRetriableError(t *testing.T) {
+	//Create sample error.
+	err := NonRetriableError(NewSampleError("Do Not Retry"))
```
Comment (jhendrixMSFT) [inline sdk/internal/errorinfo/errorinfo_test.go:15]: You could simplify this a bit by saying `errors.New("Do Not Retry")` and remove `NewSampleError` and friends.

File: sdk/internal/errorinfo/errorinfo_test.go [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+
+	// Check if the error is of type NonRetriableError
+	if _, ok := err.(*nonRetriableError); !ok {
```
Comment (richardpark-msft) [inline sdk/internal/errorinfo/errorinfo_test.go:18]: The concrete type isn't necessarily the important part - it's just an implementation detail. You'll want to check that your type conforms to the "NonRetriable" shape by using an interface.

For example:

```go
var nonRetriableErr *interface { NonRetriable() }
require.ErrorAs(err, &nonRetriableErr)
```

File: sdk/internal/errorinfo/errorinfo.go [hunk 4]
```diff
@@ -14,3 +14,24 @@ type NonRetriable interface {
 	error
 	NonRetriable()
 }
+
+// NonRetriableError marks the specified error as non-retriable.
+func NonRetriableError(err error) error {
+	return &nonRetriableError{err}
```
Comment (richardpark-msft) [inline sdk/internal/errorinfo/errorinfo.go:20]: It doesn't seem like it needs to be a pointer, btw.

File: sdk/internal/errorinfo/errorinfo_test.go (hunk 3 above)
Comment (jhendrixMSFT) [reply sdk/internal/errorinfo/errorinfo_test.go:18]: Or use the predefined interface type.
```go
var nonRetriableErr *NonRetriable
```

File: sdk/internal/errorinfo/errorinfo_test.go [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+	}
+
+	// Check error message is correct
+	errMsg := "Do Not Retry"
+	if err.Error() != errMsg {
+		t.Fatalf("Expected error message to be '%q' but got '%q'.", errMsg, err.Error())
+	}
+}
+
+func NewSampleError(message string) error {
```
Comment (richardpark-msft) [inline sdk/internal/errorinfo/errorinfo_test.go:29]: If you want there's already a convenient "create an error based on a string" with `errors.New`.

(For future stuff, if you want to format the error there's `fmt.Errorf`.)

File: sdk/internal/errorinfo/errorinfo.go (hunk 4 above)
Comment (jhendrixMSFT) [reply sdk/internal/errorinfo/errorinfo.go:20]: By convention we return errors as pointer-to-type. Plus, the receivers on `nonRetriableError` are by pointer.

File: sdk/internal/errorinfo/errorinfo_test.go (hunk 5 above)
Comment (richardpark-msft) [reply sdk/internal/errorinfo/errorinfo_test.go:29]: (if the intention here was to make sure you could "unwrap" the inner type then you'll want to do an `errors.As()` in your test too.)

File: sdk/internal/errorinfo/errorinfo.go (hunk 4 above)
Comment (MBSolomon) [reply sdk/internal/errorinfo/errorinfo.go:20]: Discussed with Richard offline. Resolving comment

File: sdk/internal/errorinfo/errorinfo.go [hunk 6]
```diff
@@ -14,3 +14,24 @@ type NonRetriable interface {
...
+}
+
+type nonRetriableError struct {
```
Comment (RickWinter) [inline sdk/internal/errorinfo/errorinfo.go:23]: Add doc comments for each of the methods.

File: sdk/internal/errorinfo/errorinfo_test.go [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+import (
+	"errors"
+	"testing"
+
+	"github.com/stretchr/testify/require"
+)
+
+func TestNonRetriableError(t *testing.T) {
+	//Create sample error.
+	err := NonRetriableError(errors.New("Do Not Retry"))
```
Comment (jhendrixMSFT) [inline sdk/internal/errorinfo/errorinfo_test.go:18]: Nit. Make `Do Not Retry` a constant so you don't have to duplicate the value.

File: sdk/internal/errorinfo/errorinfo_test.go [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+
+	// Check error message is correct
+	errMsg := "Do Not Retry"
+	if err.Error() != errMsg {
```
Comment (jhendrixMSFT) [inline sdk/internal/errorinfo/errorinfo_test.go:22]: Can simplify to `require.EqualValues(t, errMsg, err.Error())`

File: sdk/internal/errorinfo/errorinfo_test.go [hunk 9]
```diff
@@ -9,20 +9,21 @@ package errorinfo
...
-	errMsg := "Do Not Retry"
-	if err.Error() != errMsg {
-		t.Fatalf("Expected error message to be '%q' but got '%q'.", errMsg, err.Error())
+	if err.Error() != dnr {
+		t.Fatalf("Expected error message to be '%q' but got '%q'.", dnr, err.Error())
 	}
 
-	var nonRetriableErr interface{ NonRetriable() }
-	require.True(t, errors.As(&nonRetriableError{}, &nonRetriableErr))
+	var e *nonRetriableError
```
Comment (jhendrixMSFT) [inline sdk/internal/errorinfo/errorinfo_test.go:25]: This should be.
```go
var e *NonRetriable
```
This ensures that `err` implements the `NonRetriable` interface.

**Review by RickWinter — APPROVED** [review]

**Review by jhendrixMSFT — APPROVED** [review]

**Pull request merged**